YakimS/assignment2/BinaryTree.java

```diff
@@ -0,0 +1,231 @@
+package CodeUAss1Package;
+
+import javax.management.openmbean.KeyAlreadyExistsException;
+import java.util.*;
+
+/**
+ * This class contains the root to a binary tree that can have only
+ * one node for each possible key.
+ * @param <T> is the type of data stored in the tree nodes value
+ */
+public class BinaryTree<T> {
+
+    private Node root;
+    private Set<Integer> keySet;
+
+    public BinaryTree() {
+        keySet = new HashSet<>();
+    }
+
+    public void printBinaryTree()
+    {
+        printBinaryTree(this.root,0);
+    }
+    private void printBinaryTree(Node root, int level){
+        if(root==null)
+            return;
+        printBinaryTree(root.getRightSon(), level+1);
+        if(level!=0){
+            for(int i=0;i<level-1;i++)
+                System.out.print("|\t\t");
+            System.out.println("|-------"+root.getKey());
+        }
+        else
+            System.out.println(root.getKey());
+        printBinaryTree(root.getLeftSon(), level+1);
+    }
+
+    /**
+     * This method add a node to a random valid place in the binary tree
+     * @param value - value of the added node
+     * @param key - key of the added node
+     * @throws KeyAlreadyExistsException if there is already a node with key as given in the parameter
+     */
+    public void add(T value, int key) throws KeyAlreadyExistsException

```

Comment:
Key is not really needed, since it is not intended to be a map. Better to just use value as the key.